File: Kalorie/Kalorie/Core/Networking/FireStone/FoodItemSubmissionDTO.swift
//
//  FoodItemSubmissionDTO.swift
//  Kalorie
//
//  Created by Josef Antoni on 10.09.2026.
//

import Foundation

struct FoodItemSubmissionDTO: Codable {

    // MARK: - Properties

    let id: String
    let barcode: String?
    let submittedBy: String
    let status: FoodItemSubmissionStatus
    let submittedAt: TimeInterval
    let rejectReason: String?
    let item: FoodItemDTO

    // MARK: - Coding keys

    enum CodingKeys: String, CodingKey {
        case id, barcode, status, item
        case submittedBy = "submitted_by"
        case submittedAt = "submitted_at"
        case rejectReason = "reject_reason"
    }

    // MARK: - Init

    init(
        id: String,
        barcode: String?,
        submittedBy: String,
        status: FoodItemSubmissionStatus,
        submittedAt: Date,
        rejectReason: String?,
        item: FoodItemDomain
    ) {
        self.id = id
        self.barcode = barcode
        self.submittedBy = submittedBy
        self.status = status
        self.submittedAt = submittedAt.timeIntervalSince1970
        self.rejectReason = rejectReason
        self.item = FoodItemDTO(item: item)
    }

    // MARK: - Functions

    func asDomain() -> FoodItemSubmissionDomain {
        FoodItemSubmissionDomain(
            id: id,
            barcode: barcode,
            submittedBy: submittedBy,
            status: status,
            submittedAt: submittedAt.toDate,
            rejectReason: rejectReason,
            item: item.asDomain()
        )
    }
}

File: Android/app/src/main/java/antoni/kalorie/core/networking/FoodItemSubmissionDTO.kt
package antoni.kalorie.core.networking

import antoni.kalorie.core.models.FoodItemDomain
import antoni.kalorie.core.models.FoodItemSubmissionDomain
import antoni.kalorie.core.models.FoodItemSubmissionStatus
import antoni.kalorie.core.utils.epochSecondsAsDouble
import antoni.kalorie.core.utils.instantFromEpochSeconds
import java.time.Instant
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class FoodItemSubmissionDTO(
    val id: String,
    val barcode: String? = null,
    @SerialName("submitted_by") val submittedBy: String,
    val status: FoodItemSubmissionStatus,
    @SerialName("submitted_at") val submittedAt: Double,
    @SerialName("reject_reason") val rejectReason: String? = null,
    val item: FoodItemDTO,
) {

    // MARK: - Init

    constructor(
        id: String,
        barcode: String?,
        submittedBy: String,
        status: FoodItemSubmissionStatus,
        submittedAt: Instant,
        rejectReason: String?,
        item: FoodItemDomain,
    ) : this(
        id = id,
        barcode = barcode,
        submittedBy = submittedBy,
        status = status,
        submittedAt = submittedAt.epochSecondsAsDouble(),
        rejectReason = rejectReason,
        item = FoodItemDTO(item),
    )

    // MARK: - Functions

    fun asDomain(): FoodItemSubmissionDomain = FoodItemSubmissionDomain(
        id = id,
        barcode = barcode,
        submittedBy = submittedBy,
        status = status,
        submittedAt = instantFromEpochSeconds(submittedAt),
        rejectReason = rejectReason,
        item = item.asDomain(),
    )
}
